### ISSUE TITLE
dynamic branch tracking

### ISSUE THREAD

**Issue by tfhartmann** [puppetlabs/r10k#23]
I'm not sure exactly how to phrase this feature request, but here goes.  It would be cool if you were able to set a reference to track a branch in the requested repo that matched the branch that r10k was checking out. 

So for example if setting something like this, in the "development" branch would track a branch called "development" in puppet-filemapper.   If the environment directory is master, then track master in the puppet-filemapper repo, and so on. 

mod 'filemapper',
  :git => 'git://github.com/adrienthebo/puppet-filemapper.git',
  :ref => 'environment'

**Comment by adrienthebo** [comment]
Sorry for the slow response; I'm trying to get 1.0.0 out the door and haven't had the time to handle tickets.

I've been interested in something like this, but I haven't been sure about the best implementation. So given something like `:ref => 'environment'` or `:track => true`, how should this probe? If there's a remote branch, then try it? In other case, default to master?

**Comment by tfhartmann** [comment]
Yes, something like that. My through was that say I create a feature branch in puppet called 'tfhartmann-feature' and I wanted to deploy a module that I was working on with r10k, but I don't want it to track the stable branch, I want it to track the branch that matches my new puppet environment, but in production I'll want it to track stable, i.e. master/production (I use master as production)  so the tfhartmann-feature environment would track the "tfhartmann-feature" branch in the repo that r10k deploys.  So master would track repo/master development would track repo/development and tfhartmann-feature would track tfhartmann-feature, and if $branchname didn't exist default to master.  So whatever $environment i'm in, I'd also track a branch called $environment if is available.  

With the exception of master maybe, make master == production, I'm not sure how other folks are doing this with pupet,  I ended up hardcoding master == production in puppet.conf

[production]
    manifest   = $confdir/environments/master/manifests/site.pp
    modulepath = $confdir/modules:$confdir/environments/master/modules:$confdir/environments/master/dist:$confdir/environments/master/site

**Comment by adrienthebo** [comment]
Targeting this feature at `1.3.0`

**Comment by tfhartmann** [comment]
:+1:

**Comment by antaflos** [comment]
I think I finally understand this feature request and believe it would be incredibly useful to have r10k checkout a module's branch based on the branch of the Puppetfile (= environment) that references the module.

This way, if I understand correctly, one could easily implement more dynamic environments where multiple admins could work independently of each other in their own environment. So a Puppetfile in branch "antaflos-testing" would make r10k create an environment "antaflos-testing" and all modules in that Puppetfile would be checked out to "antaflos-testing", if such a branch exists for that module. I could then work on a module's feature, push that feature to "antaflos-testing" and test it on a Puppet node that has `environment = antaflos-testing`. Another admin could, in this vein, work on the same module on a different feature in a different branch and test that module on a node that has `environment = otheradmin-testing`.

Maybe r10k could by default try to checkout a module's branch based on the Puppetfile's branch (= environment) and if no such branch exists, default to "master". It should of course still be possible to explicitly set a `:ref => ...`.

**Comment by krzyzan** [comment]
Definitely useful. It would reduce number of steps to create feature branch using dynamic environments and r10k. Right now people do some automated Puppetfile search/replace to accomplish this: http://philzim.com/2014/05/02/an-automated-r10k-workflow-that-works/

Something like this would do:
`track => true` (use module branch with the same name as Puppetfile repo branch)
`ref => master` (default if module branch not present)

**Comment by vchepkov** [comment]
For what it worth I am highly interested in this feature as well

**Comment by etlweather** [comment]
I'm actually trying to do exactly that. Currently, I only need a much simpler functionality however. For one module, which is being pulled from Git, I want to make it download the branch that match the environment.

I read in the doc on the Puppetfile that being Ruby DSL, any Ruby code should work there. 

> Because the Puppetfile format is actually implemented using a Ruby DSL any valid Ruby expression can be used.

Unfortunately, I'm not a big Ruby guy and I have no idea what variables would be available (or in what form).

Basically, right now I'm trying to do something like this in my Puppetfile:

```
mod 'app',
  :git    => 'https://github.com/apps/app.git',
  :branch => ${environment}
```

which happens to be just what was commented here: https://github.com/puppetlabs/r10k/pull/101#issuecomment-58616196

**Comment by etlweather** [comment]
@vchepkov - did you ever figured out how to do what you wrote about in https://github.com/puppetlabs/r10k/pull/101#issuecomment-58616196 ?

The 

```
mod 'app',
  :git    => 'https://github.com/apps/app.git',
  :branch => ${environment}
```

idea?

**Comment by vchepkov** [comment]
I found a workaround. Not really something I like, but it works until a proper solution is implemented in r10k. I used r10k multi-tenancy feature to deploy branches form all application's repositories:

```
r10k::sources:
  main:
    remote: https://github.com/apps/main.git
    basedir: /etc/puppetlabs/puppet/environments
    invalid_branches: correct
  app1:
    remote: https://github.com/apps/app1.git
    basedir: /etc/puppetlabs/puppet/modules/app1
    invalid_branches: correct
  app2:
    remote: https://github.com/apps/app2.git
    basedir: /etc/puppetlabs/puppet/modules/app2
    invalid_branches: correct
  r10k::purgedirs:
  - /etc/puppetlabs/puppet/environments
  - /etc/puppetlabs/puppet/modules/app1
  - /etc/puppetlabs/puppet/modules/app2
```

then in main repository environments.conf I listed them all in modulepath:

```
modulepath = modules:site:$confdir/modules/app1/$environment:$confdir/modules/apps2/$environment:$basemodulepath
```

As I said, not pretty, but this works in PE 3.3. For unknown reasons PuppetLabs decided to stop interpolating $environment in modulepath of environment.conf (and only in modulepath, mind you, so it was a deliberate "crippling") in PE 3.7. And this is one of the many reasons we will stay with PE3.3 for foreseeable future

**Comment by etlweather** [comment]
@vchepkov  - Figured out a solution:

```
mod 'app',
  :git    => 'https://github.com/apps/app.git',
  :branch => @librarian.basedir.split('/').last
```

That sets the branch to the environment.

**Comment by vchepkov** [comment]
It is a big step in the right direction, thanks.
The only problem, r10k fais each time while deploying "feature" branches in "main" repo, since these branches are missing in "app" repo: 

```
[R10K::TaskRunner - ERROR] Task #<R10K::Task::Environment::Deploy:0x00000002dd37e8> failed while running: Cannot check out unresolvable ref
```

I think a ruby script needs to be called instead and return some 'default' branch if not a "supported" environment is returned by @librerian call

**Comment by etlweather** [comment]
Yeah, that could be a problem. However for my use right now that is all I needed. But when I was testing around, I did pass a branch name that wasn't existing and what I remember R10K doing on those was checking out the master branch instead. I have version 1.4.0.

**Comment by vchepkov** [comment]
I upgraded to 1.4.0, it still throws an error though:

```
[R10K::Action::Deploy::Environment - ERROR] Could not locate hash at /etc/puppetlabs/puppet/environments/feature_puppetfile/modules/test
```

And it seems code you suggested doesn't handle branches with '/' in the name, like feature/puppetfile in my case.

**Comment by elyscape** [comment]
@vchepkov Yeah, that's a bug with 1.4.0.  It was fixed in 1.4.1 as per #260.

**Comment by vchepkov** [comment]
@elyscape, That's not what I was saying.
I don't know what @librarian.basedir function returns, but I bet for branch 'feature/puppetfile' @librarian.basedir.split('/').last will return 'puppetfile', not 'feature/puppetfile'

**Comment by elyscape** [comment]
Ahh, gotcha.  Yeah, that makes sense.
